pooch/downloaders.py

```diff
@@ -804,29 +888,33 @@ def download_url(self, file_name):
             The HTTP URL that can be used to download the file.
         """
 
-        # Use the figshare API to find the article ID from the DOI
-        article = requests.get(
-            f"https://api.figshare.com/v2/articles?doi={self.doi}"
-        ).json()[0]
-        article_id = article["id"]
-        # With the ID, we can get a list of files and their download links
-        response = requests.get(
-            f"https://api.figshare.com/v2/articles/{article_id}/files"
-        )
-        response.raise_for_status()
-        files = {item["name"]: item for item in response.json()}
+        files = {item["name"]: item for item in self.api_response}
         if file_name not in files:
             raise ValueError(
                 f"File '{file_name}' not found in data archive {self.archive_url} (doi:{self.doi})."
             )
         download_url = files[file_name]["download_url"]
         return download_url
 
+    def populate_registry(self, pooch):
+        """
+        Populate the registry using the data repository's API
+
+        Parameters
+        ----------
+        pooch : Pooch
+            The pooch instance that the registry will be added to.
+        """
+
+        for filedata in self.api_response:
+            pooch.registry[filedata["name"]] = f"md5:{filedata['computed_md5']}"
+
 
 class DataverseRepository(DataRepository):  # pylint: disable=missing-class-docstring
-    def __init__(self, doi, archive_url):
+    def __init__(self, doi, archive_url, response):
```

Comment:
Thanks a lot for the detailed feedback. I do agree that your design is better and keeps the public API of the `DataverseRepository` class intact. I have implemented it with the following modifications:
* I used `@classmethod` over `@staticmethod` as it allows `self.foo()` instead of `type(self).foo()`
* I prefixed `get_api_response` with an underscore, as I do not consider it a public API. The corresponding public API is simply using the `api_response` property in absence of a cached result